### PR TITLE
ISPN-11807 Create a FORKED mode for InfinispanServerRule

### PR DESCRIPTION
**Pull request by gustavolira**
https://issues.redhat.com/browse/ISPN-11807

https://github.com/gustavolira/infinispan-playground-testdriver (a fork of Tristan's work) which demonstrates how to use this in other projects.

### CODE REVIEW

**Comment by karesti** [comment]
would be nice to be implemented in the infinispanserverextension as well (junit5)

**Comment by diegolovison** [comment]
Why not implement Forked mode on top of Arquillian?
We were using https://github.com/infinispan/infinispan-arquillian-container

**Comment by gustavolira** [comment]
> https://github.com/infinispan/infinispan-arquillian-container

Because the intention is to use exactly as InfinispanServerRule, letting the rule create the caches has the following advantages:

- preconfigures the clients against the servers
- automatic cache creation scoped to the method
- automatic cache removal after method execution
- unique naming of caches so that different test methods don't attempt to use the same
- the generated cache names can be used as database identifiers without the risk of going over size limits (e.g. MySQL has a 64-character limit on table/column names, 63 for PostgreSQL)

**Comment by gustavolira** [comment]
> rebase and junit5

Rebased and working with junit5

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    @Override
+    public InetAddress getServerAddress(int server) {
+        return Exceptions.unchecked(() -> InetAddress.getByName("localhost"));
+    }
+
+    @Override
+    public void pause(int server) {
+        throw new UnsupportedOperationException();
```
Comment (tristantarrant) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java:102]: Why not `kill -STOP`

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    @Override
+    public void resume(int server) {
```
Comment (tristantarrant) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java:106]: and `kill -CONT`

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedServer.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        try {
+            process = pb.start();
+            isServerStarted = runWithTimeout(this::checkServerLog, START_PATTERN);
+            if (!isServerStarted) {
+                throw new IllegalStateException("The server couldn't start");
+            }
+        } catch (Exception e) {
+            log.error(e);
+        }
+        log.info("SERVER STARTED!");
```
Comment (tristantarrant) [inline server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedServer.java:82]: No need for this.

File: server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java (hunk 1 above)
Comment (gustavolira) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java:102]: @tristantarrant IMO with a forked process makes more sense to totally stop the process, release the port opened and if needed create a new process.

Comment (diegolovison) [reply server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java:102]: IMO, the forked mode shouldn't support pause or continue.
`EmbeddedInfinispanServerDriver` doesn't support `pause` or `continue`.
If we need to deal with those scenarios we must use `ContainerInfinispanServerDriver`
The `throw new UnsupportedOperationException();` looks good to me.

File: server/testdriver/core/pom.xml [hunk 4]
```diff
@@ -73,5 +73,18 @@
...
       </dependency>
    </dependencies>
+
+   <build>
+      <plugins>
+         <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-compiler-plugin</artifactId>
+            <configuration>
+               <release>11</release>
```
Comment (tristantarrant) [inline server/testdriver/core/pom.xml:83]: why is this needed ?

Comment (gustavolira) [reply server/testdriver/core/pom.xml:83]: Without that I can't use `process.pid`, I'm explicitly setting JDK 11 as compiler

Comment (diegolovison) [reply server/testdriver/core/pom.xml:83]: Forked mode shouldn't support kill, stop or cont. Please remove `process.pid`. If you wish to `stop` the server. Call the stop endpoint from the `RestClient`

Comment (tristantarrant) [reply server/testdriver/core/pom.xml:83]: Ah, you need to have the PID of the process

Comment (gustavolira) [reply server/testdriver/core/pom.xml:83]: > Forked mode shouldn't support kill, stop or cont. Please remove `process.pid`. If you wish to `stop` the server. Call the stop endpoint from the `RestClient`

@diegolovison I'm stopping the server with RestClient already, you can see here https://github.com/infinispan/infinispan/blob/704ef6d1b727a7ca79a462290fed703879467b5c/server/testdriver/core/src/main/java/org/infinispan/server/test/core/ForkedInfinispanServerDriver.java#L69

Comment (diegolovison) [reply server/testdriver/core/pom.xml:83]: We still support Java 8 to run the tests.
What will be the behavior when we enable the profile `java8-test` when running the tests?

If we are going to support `kill -9` why can't we support `kill -STOP` or `kill -CONT`? IMHO, I do prefer have those kinds of tests with `ContainerInfinispanServerDriver`

**Comment by tristantarrant** [comment]
@gustavolira I have pushed quite a few changes:
* removed the Java 11 requirement and used the pid support in the infinispan-commons jdkspecific module
* removed the unused getLog() methods
* removed the property that allows to override the configuration set by the test: none of the other drivers allow this, so I don't think the FORKED driver should be special in this respect
* Put some sanity in the TestSystemPropertyNames
* formatted everything with the Infinispan code style (you should import those into your IDE)

File: server/testdriver/core/pom.xml (hunk 4 above)
Comment (alanfx) [reply server/testdriver/core/pom.xml]: > Forked mode shouldn't support kill, stop or cont. Please remove `process.pid`. If you wish to `stop` the server. Call the stop endpoint from the `RestClient`

I disagree with this. I see no point in hobbling `FORKED` mode. It should have the same capabilities as the other modes of `InfinispanServerRule`.

BTW, ping @rhusar you'll be interested in this PR

**Comment by rhusar** [comment]
@gustavolira BTW this needs a rebase now.

File: server/testdriver/core/pom.xml (hunk 4 above)
Comment (rhusar) [reply server/testdriver/core/pom.xml]: @alanfx @gustavolira Yes - very interested - thanks for the ping. Will take it for a spin later this week.

Comment (rhusar) [reply server/testdriver/core/pom.xml]: @diegolovison 

>Forked mode shouldn't support kill

I am not following why, could you explain? Actually, FORKED mode is kind of ideal for supporting a JVM kill.

**Comment by gustavolira** [comment]
> @gustavolira BTW this needs a rebase now.

 @ryanemerson @rhusar rebased

File: commons/jdk11/src/main/java/org/infinispan/commons/jdkspecific/ProcessInfo.java [hunk 5]
```diff
@@ -24,6 +24,10 @@ public static ProcessInfo getInstance() {
       return new ProcessInfo(ProcessHandle.current());
    }
 
+   public static Process of(java.lang.Process process) {
+      return new Process(process.toHandle());
```
Comment (rhusar) [inline commons/jdk11/src/main/java/org/infinispan/commons/jdkspecific/ProcessInfo.java:28]: Failure from CI:

`17:24:33  13:22:27.631 [ERROR] /home/infinispan/workspace/Infinispan_PR-8297/commons/jdk11/src/main/java/org/infinispan/commons/jdkspecific/ProcessInfo.java:[28,13] error: Process is abstract; cannot be instantiated`

Comment (rhusar) [reply commons/jdk11/src/main/java/org/infinispan/commons/jdkspecific/ProcessInfo.java:28]: Perhaps this was meant to be:

```
-   public static Process of(java.lang.Process process) {
-      return new Process(process.toHandle());
+   public static ProcessInfo of(Process process) {
+      return new ProcessInfo(process.toHandle());
```

Comment (gustavolira) [reply commons/jdk11/src/main/java/org/infinispan/commons/jdkspecific/ProcessInfo.java:28]: @tristantarrant can you please take a look?

**Comment by tristantarrant** [comment]
I've made a few changes and merged to master.

File: server/testdriver/core/pom.xml (hunk 4 above)
Comment (diegolovison) [reply server/testdriver/core/pom.xml]: @rhusar, when you start multiple processes on the same OS and then start doing some chaos testing like ( kill, stop or cont ) those tests will start failing sometimes and you won't be able to reproduce it locally ( flaky tests ). The best option is to use Containers for those kinds of tests or isolates in different VMs/Machines.

This is why for me it is a `-1` to support kill, stop or cont.

**Comment by tristantarrant** [comment]
I actually implemented pause/resume. Use at your own risk.
